### PR TITLE
[ALLUXIO-1718] Delete block metadata when a file is deleted.

### PR DESCRIPTION
**Pull request by ifcharming**
https://tachyon.atlassian.net/browse/ALLUXIO-1718

Currently the block metadata is not deleted when a file is deleted, and therefore block master regards the blocks as lost. This handling slows the system down, because the lost block will trigger lost file detection, which will in turn lead to a FileNotFound exception in inodetree look up.

Therefore, we should distinguish between freeing and deleting files in Alluxio.

### CODE REVIEW

**Comment by ifcharming** [comment]
@yupeng9 @gpang

**Comment by yupeng9** [comment]
LGTM. Thanks for the bug fix!

File: core/server/src/main/java/alluxio/master/block/BlockMaster.java [hunk 1]
```diff
@@ -305,8 +305,9 @@ public long getUsedBytes() {
    * Removes blocks from workers.
    *
    * @param blockIds a list of block ids to remove from Alluxio space
+   * @param deleteBlocksMetadata whether to delete blocks metadata in Master
    */
-  public void removeBlocks(List<Long> blockIds) {
+  public void removeBlocks(List<Long> blockIds, boolean deleteBlocksMetadata) {
```
Comment (gpang) [inline core/server/src/main/java/alluxio/master/block/BlockMaster.java]: NIT: I would prefer the name to be "delete"

File: core/server/src/main/java/alluxio/master/block/BlockMaster.java [hunk 2]
```diff
@@ -322,6 +323,9 @@ public void removeBlocks(List<Long> blockIds) {
             }
           }
           mLostBlocks.remove(blockId);
```
Comment (gpang) [inline core/server/src/main/java/alluxio/master/block/BlockMaster.java]: I have a question about this. Why doesn't this remove the block from being considered as lost? Shouldn't this line take care of that?

**Comment by gpang** [comment]
I left a few small questions/comments. Thanks!

File: core/server/src/main/java/alluxio/master/block/BlockMaster.java (hunk 1 above)
Comment (ifcharming) [inline core/server/src/main/java/alluxio/master/block/BlockMaster.java]: Done.

File: core/server/src/main/java/alluxio/master/block/BlockMaster.java (hunk 2 above)
Comment (ifcharming) [inline core/server/src/main/java/alluxio/master/block/BlockMaster.java]: Good catch. I think you're right:
- for Free, blocks should be added to mLostBlocks
- for delete, blocks should be deleted from both mBlocks and mLostBlocks

I have a follow-up question:
Can one block be shared by multiple files?

**Comment by ifcharming** [comment]
Thanks @gpang, updated, can you PTAL?

File: core/server/src/main/java/alluxio/master/block/BlockMaster.java (hunk 2 above)
Comment (gpang) [inline core/server/src/main/java/alluxio/master/block/BlockMaster.java]: Any block belongs to only 1 file.

**Comment by ifcharming** [comment]
Updated and added comment for the different logic between free and delete.
PTAL.

Will add integration tests for this in a following PR.

**Comment by ifcharming** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
LGTM

**Pull request merged**